### ISSUE TITLE
ioRedis can't delete key in redis cluster 

### ISSUE THREAD

**Issue by KurooUsagy** [luin/ioredis#914]
**Configurations** 

ioRedis@4.9.0 (cluster mode)
connect-redis@3.4.1
express-session@1.15.6
Redis Server Version: 5

**Errors** 

> ReplyErrorredis-parser.lib:parser in parseError
> errorCROSSSLOT Keys in request don't hash to the same slot

the Reply errors contains : 

`{
command: {
args: [
sess:6BSHHdpoqDVhpqWuOqrunZcqsUadtSEn, 
], 
name: del
}
}`

**My code** 

```
const express = require('express');
const session = require('express-session');
const Redis = require('ioredis');

redisClient = new Redis.Cluster([{ 'host': host, 'port': port }]); 
const RedisStore = require('connect-redis')(session);

let sessionStore = new RedisStore({
    client: redisClient
});

app.use(session({
    secret: 'mysecret',
    saveUninitialized: false,
    resave: false,
    store: sessionStore,
    cookie: {
        maxAge: 60 * 1000 // 1 minutes 
    }
}));
```

**Logs**

> connect:redis DEL "Nu-53UpCkkG4XutUUFULWlWH_s-M2IU9"
> ioredis:redis write command[10.2.0.29:6379]: 0 -> del([ 'sess:Nu-53UpCkkG4XutUUFULWlWH_s-M2IU9', '' ])

After seeing that I suppose that ioredis add a blank parameters and this is why it can't delete my session key.

**What is expected ?**

I expect that my session key will be deleted in my redis cluster.

Thank in advance for your reply 😃

**Comment by luin** [comment]
The issue is caused by https://github.com/tj/connect-redis/blob/v3.4.1/lib/connect-redis.js#L229.

When calling `req.session.destroy()` without providing a callback function, connect-redis will actually invoke `redis.del(sid, undefined)`. ioredis considers `undefined` as empty strings, which makes the invocation become `redis.del(sid, "")`.

The solutions are:
1. Create a pull request for connect-redis that check if the callback function is provided: `fn ? this.client.del(sid, fn) : this.client.del(sid);`;
2. Make ioredis ignore trailing `undefined`s. It will be released in v5.0 (within about three months) since this is a breaking change.

**Comment by armenr** [comment]
We've hit the same issue. This can be very bad, depending on what the TTL is for your session objects in Redis.

We went around this with the following:

```
const session = require('express-session');
const RedisStore = require('connect-redis')(session);
const Redis = require('ioredis');
const redisClient = new Redis.Cluster([{ port: REDIS_PORT, host: REDIS_HOST }]);
```

```
router.use(
    session({
      name: SESSIONS_KEY,
      genid: () => uuidv4(),
      store: new RedisStore({
        client: redisClient,
        prefix: SESSIONS_KEY,
        ttl: Number(SESSION_EXPIRATION_TIME_MS / 1000),
      }),
      secret: SESSION_SECRET,
      resave: false,
      saveUninitialized: true,
      cookie: {
        httpOnly: true,
        domain: COOKIE_DOMAIN,
        maxAge: Number(SESSION_EXPIRATION_TIME_MS),
      },
    }),
  );
```


```
  router.post('/logout', (req, res) => {
    try {
      res.clearCookie(SESSIONS_KEY);
      redisClient.del(SESSIONS_KEY + req.signedCookies[SESSIONS_KEY]);
      res.send();
    } catch (e) {
      console.log('LOGOUT SCREWED THE POOCH!', e);
      res.status(401).send();
    }
  });

```

**Comment by luin** [comment]
I've created a pull request on the connect-redis repo for this issue: https://github.com/tj/connect-redis/pull/272

**Comment by armenr** [comment]
Thanks for getting on top of this! :)

Not a big issue...just a silently unexpected one. 👍

**Comment by luin** [comment]
Fixed in connect-redis@v3.4.2.
